File: sdks/csharp/src/Verdocs.Sdk/Utils/Entitlements.cs
using Verdocs.Models;

namespace Verdocs.Utils;

/// <summary>Entitlement collapsing (js-sdk: Utils/Entitlements.ts).</summary>
public static class Entitlements
{
    /// <summary>
    /// Collapses raw entitlement records down to the active one per feature. Only entries
    /// whose date window covers now survive, and the first record per feature wins, so
    /// presence of a key means the feature is currently enabled. This is the client-side twin
    /// of what <see cref="Resources.Organizations.GetActiveEntitlementsAsync"/> returns.
    /// </summary>
    /// <param name="entitlements">
    /// The raw records, e.g. from <see cref="Resources.Organizations.GetEntitlementsAsync"/>.
    /// </param>
    /// <returns>Feature name (see <see cref="EntitlementFeature"/>) to the entitlement record currently granting it.</returns>
    public static IReadOnlyDictionary<string, Entitlement> CollapseEntitlements(IEnumerable<Entitlement> entitlements)
    {
        ArgumentNullException.ThrowIfNull(entitlements);

        // Grants may overlap, so the first grant whose date window covers now wins for each
        // feature.
        var now = DateTimeOffset.UtcNow;
        var active = new Dictionary<string, Entitlement>();
        foreach (var entitlement in entitlements)
        {
            if (now >= entitlement.StartsAt && now <= entitlement.EndsAt)
            {
                active.TryAdd(entitlement.Feature, entitlement);
            }
        }

        return active;
    }
}

File: sdks/python/src/verdocs/utils/entitlements.py
"""Entitlement collapsing (js-sdk: Utils/Entitlements.ts)."""

from __future__ import annotations

from datetime import datetime, timezone

from ..models.core import Entitlement
from ..models.organizations import ActiveEntitlements


def collapse_entitlements(entitlements: list[Entitlement]) -> ActiveEntitlements:
    """Collapse raw entitlement records down to the active one per feature.

    Only entries whose date window covers now survive, and the first record
    per feature wins, so presence of a key means the feature is currently
    enabled. This is the client-side twin of what
    organizations.get_active_entitlements() returns.

    Args:
        entitlements: The raw records, e.g. from organizations.get_entitlements().

    Returns:
        Feature name to the entitlement record currently granting it.

    @sdkOperation entitlement.collapseEntitlements
    @sdkGroup Entitlement
    @sdkPage Helpers
    """
    now = datetime.now(timezone.utc)
    active: ActiveEntitlements = {}
    for entitlement in entitlements:
        if entitlement.starts_at <= now <= entitlement.ends_at and entitlement.feature not in active:
            active[entitlement.feature] = entitlement
    return active
